File: scripts/python/src/tools/dotfile/profiles.py
import os
import shutil

KDE_COMMANDS = ("plasmashell", "startplasma-wayland", "startplasma-x11")
HYPRLAND_COMMANDS = ("Hyprland", "hyprctl")

DESKTOP_GROUPS = {
    "linux/kde": "kde",
    "linux/hyprland": "hyprland",
}


def list_profiles(envdir):
    found = []
    for parent, _dirnames, filenames in os.walk(envdir):
        if "manifest" in filenames:
            found.append(os.path.relpath(parent, envdir))
    return sorted(found)


def read_os_release(path="/etc/os-release"):
    values = {}
    try:
        with open(path, encoding="utf-8") as handle:
            lines = handle.read().splitlines()
    except OSError:
        return values
    for line in lines:
        if "=" not in line:
            continue
        key, _, value = line.partition("=")
        value = value.strip('"').strip("'")
        values[key] = value
    return values


def detect_linux_platform(os_release=None):
    values = read_os_release() if os_release is None else os_release
    platform_id = values.get("ID", "")
    id_like = values.get("ID_LIKE", "")
    if platform_id == "arch":
        return "arch-linux"
    if platform_id == "ubuntu":
        return "ubuntu"
    likes = id_like.split()
    if "arch" in likes:
        return "arch-linux"
    if "ubuntu" in likes:
        return "ubuntu"
    return ""


def detect_platform():
    system = os.uname().sysname
    if system == "Darwin":
        return "macos"
    if system == "Linux":
        return detect_linux_platform()
    return ""


def detect_installed_desktops():
    desktops = []
    if any(shutil.which(command) for command in KDE_COMMANDS):
        desktops.append("kde")
    if any(shutil.which(command) for command in HYPRLAND_COMMANDS):
        desktops.append("hyprland")
    return desktops


def manifest_group_lines(manifest):
    groups = []
    with open(manifest, encoding="utf-8") as handle:
        for line in handle.read().splitlines():
            group = line.split("#", 1)[0].strip()
            if group:
                groups.append(group)
    return groups


def profile_matches_host(envdir, profile, platform, desktops):
    if platform == "macos":
        if profile != "macos":
            return False
    elif profile.split("/")[0] != platform:
        return False
    manifest = os.path.join(envdir, profile, "manifest")
    for group in manifest_group_lines(manifest):
        required = DESKTOP_GROUPS.get(group)
        if required and required not in desktops:
            return False
    return True


def filter_profiles(envdir, platform, desktops):
    return [
        profile
        for profile in list_profiles(envdir)
        if profile_matches_host(envdir, profile, platform, desktops)
    ]


def list_relevant_profiles(envdir):
    platform = detect_platform()
    if not platform:
        return []
    return filter_profiles(envdir, platform, detect_installed_desktops())


def normalize_profile_arg(value):
    if value.startswith("--") and len(value) > 2:
        return value[2:]
    return value

File: scripts/rust/crates/dotfile/src/config/profiles.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use crate::context::Context;

pub fn command_path(context: &Context, name: &str) -> Option<PathBuf> {
    if name.contains('/') {
        let path = PathBuf::from(name);
        return executable(&path).then_some(path);
    }
    std::env::split_paths(&context.env("PATH").unwrap_or_default())
        .map(|directory| directory.join(name))
        .find(|path| executable(path))
}

fn executable(path: &Path) -> bool {
    let Ok(metadata) = fs::metadata(path) else {
        return false;
    };
    if !metadata.is_file() {
        return false;
    }
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        metadata.permissions().mode() & 0o111 != 0
    }
    #[cfg(not(unix))]
    {
        true
    }
}

pub fn linux_platform(text: &str) -> &str {
    let values = text
        .lines()
        .filter_map(|line| line.split_once('='))
        .map(|(key, value)| (key, value.trim_matches(['\'', '"'])))
        .collect::<BTreeMap<_, _>>();
    match values.get("ID").copied().unwrap_or_default() {
        "arch" => "arch-linux",
        "ubuntu" => "ubuntu",
        _ => {
            let likes = values.get("ID_LIKE").copied().unwrap_or_default();
            if likes.split_whitespace().any(|name| name == "arch") {
                "arch-linux"
            } else if likes.split_whitespace().any(|name| name == "ubuntu") {
                "ubuntu"
            } else {
                ""
            }
        }
    }
}

pub fn platform() -> String {
    match std::env::consts::OS {
        "macos" => "macos".into(),
        "linux" => {
            linux_platform(&fs::read_to_string("/etc/os-release").unwrap_or_default()).into()
        }
        _ => String::new(),
    }
}

pub fn relevant(context: &Context) -> Result<Vec<String>, String> {
    let desktop = [
        (
            "linux/kde",
            ["plasmashell", "startplasma-wayland", "startplasma-x11"]
                .iter()
                .any(|name| command_path(context, name).is_some()),
        ),
        (
            "linux/hyprland",
            ["Hyprland", "hyprctl"]
                .iter()
                .any(|name| command_path(context, name).is_some()),
        ),
    ];
    filter(context, &platform(), &desktop)
}

pub fn filter(
    context: &Context,
    platform: &str,
    desktops: &[(&str, bool)],
) -> Result<Vec<String>, String> {
    let mut found = Vec::new();
    if platform.is_empty() {
        return Ok(found);
    }
    for profile in context.profiles()? {
        if (platform == "macos" && profile != "macos")
            || profile.split('/').next() != Some(platform)
        {
            continue;
        }
        let groups = super::read_manifest(&context.manifest(&profile))?;
        if desktops
            .iter()
            .all(|(group, installed)| *installed || !groups.iter().any(|entry| entry == group))
        {
            found.push(profile);
        }
    }
    Ok(found)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn linux_distribution_and_derivative_detection() {
        for (input, expected) in [
            ("ID=arch\n", "arch-linux"),
            ("ID=ubuntu\n", "ubuntu"),
            ("ID=cachyos\nID_LIKE=arch\n", "arch-linux"),
            ("ID=neon\nID_LIKE=\"ubuntu debian\"\n", "ubuntu"),
            ("ID=gentoo\n", ""),
        ] {
            assert_eq!(linux_platform(input), expected);
        }
    }
}
